waspc/src/Wasp/Generator/DbGenerator/Jobs.hs

```diff
@@ -27,28 +40,105 @@ migrateDev projectDir = do
+  let serverDir = projectDir </> serverRootDirInProjectRootDir
+  runNodeCommandAsJob serverDir "npm" ["install"] J.Db
+
+data JobMessageForwardingStrategy = ForwardEverything | ForwardOnlyRetryErrors
+
+data Attempt = InitialAttempt | Retry
+
+-- | Runs the original job, and if there is an error, it attempts recovery by running the second job
+-- before retrying the original one last time.
+-- NOTE(shayne): We only want to forward one JobExit message to the external channel from this function, as callers will stop
+-- listening once they get it. This is why our job message forwarders do not send them and we do ourselves at the end.
+retryJobOnErrorWith :: J.Job -> J.Job -> JobMessageForwardingStrategy -> J.Job
+retryJobOnErrorWith originalJob afterErrorJob forwardingStrategy externalChannel = do
+  let channelForwarder = case forwardingStrategy of
+        ForwardEverything -> allJobOutput
+        ForwardOnlyRetryErrors -> silentUntilRetryFails
+  intermediateChannel <- newChan
+
+  (initialExitCode, _) <- concurrently (originalJob intermediateChannel) (channelForwarder InitialAttempt intermediateChannel externalChannel)
+  exitCode <-
+    case initialExitCode of
+      ExitSuccess -> return ExitSuccess
+      ExitFailure _ -> do
+        -- NOTE(shayne): We don't care what the exit code of the recovery job is. If it succeeds, great, and hopefully retry works.
+        -- If it fails, retry will likely fail again in the same manner. There may be other quirks but this is
+        -- sufficient for the npx and npm install use case we are targeting now.
+        _ <- concurrently (afterErrorJob intermediateChannel) (channelForwarder InitialAttempt intermediateChannel externalChannel)
+        (retryExitCode, _) <- concurrently (originalJob intermediateChannel) (channelForwarder Retry intermediateChannel externalChannel)
+        return retryExitCode
+
+  writeJobExitChanMessage exitCode
+  return exitCode
+  where
+    writeJobExitChanMessage exitCode =
+      writeChan
+        externalChannel
+        J.JobMessage
+          { J._data = J.JobExit exitCode,
+            J._jobType = J.Db
+          }
+
+    -- Forwards all JobOutput messages to the output channel, but does not send JobExit
+    -- as multiple jobs may be forwarded to the same output channel. The caller is responsible
+    -- for sending a final JobExit.
+    allJobOutput :: Attempt -> Chan JobMessage -> Chan JobMessage -> IO ()
+    allJobOutput attempt fromChan toChan = do
+      message <- readChan fromChan
+      case J._data message of
+        JobOutput _ _ -> writeChan toChan message >> allJobOutput attempt fromChan toChan
+        JobExit _ -> return ()
+
+    -- Only forwards JobOutput messages for the retry attempt if it fails.
+    -- All first attempt output is ignored, and so is success on retry.
+    silentUntilRetryFails :: Attempt -> Chan JobMessage -> Chan JobMessage -> IO ()
+    silentUntilRetryFails attempt fromChan toChan = silentUntilRetryFails' []
```

Comment:
FYI, when defining inner functions like `silentUntilRetryFails'` that do nothing but are a bit more general and then you call them with some initial args, there is a convention in Haskell to name them `go`. That said, I don't think it is a very good convention, so this is probably better what you did.